README.md
# WebChat
Making a web chat to train with Node.js &amp; Socket.io

## Requirements
- Git
- Node.js >= LTS 10.15.3

## Install
```bash
git clone https://github.com/Noxilex/WebChat
cd WebChat
npm install
```

## Start
```bash
npm start
```
- Open browser on : `localhost:5001`

## TODO 

- [ ] Fix chat receiving messages even when not connected to chat
- [X] Fix 2 connexions created when clicking multiple times on join button
- [X] Fix time missing leading 0
- [X] Add random color attribution for connected user
- [X] Commands
    - [X] /help
    - [X] /rename
    - [X] /roll
    - [X] /color
- [ ] Emojis interface
- [ ] Don't autoscroll when reading above
- [ ] Refactor code to make it cleaner
- [ ] Ping someone on @Username
- [ ] Animations on click, onmessage
- [ ] Timeout before sending new message
- [ ] Auto-reconnect after DC
- [X] Don't send message on empty/space message
- [ ] Scroll all the way to the bottom when you focus chat (phones)

## Suggestion

- Profile picture
- Games
    - Quizz (Game of Thrones)
    - 
- Anime music


CHANGELOG.md
# Changelog

## [0.0.4] - 2019-05-24

### Added

- Outline on /color message info
- /help command

### Changed

- Removed focus on chatTextArea after join
- Now shows date on mobile when we press the message

### Fixed

- Can't send empty messages with spaces
- Empty username are no longer added to dom (taking small space in front of message)
- Black is base color for chat history
- User-select in chat history

### Removed

### Security


## [0.0.3] - 2019-05-23

### Added

- Can rename to a multiple word name
- /roll command
- Error message in chat when user lose connexion to server
- Feedback to user when command is executed
- Command color support
- Added colors to the usernames

### Changed

- Changed username color style

### Fixed

- Fix 2 connexions created when clicking multiple times on join button
- Empty username
- Fixed hour to print 00:00 format (previous version didn't have leading 0s)
- Added max limit to the number of messages stored on the server messages array, prevents server memory overload

## [0.0.2] - 2019-05-22

### Added
- Join when you press enter on the input
- Validation on join input (between 3 & 20 characters)
- Minimum size of 1 to send message
- Maximum size of 255 to send message (Twitter size)

### Changed
- Made responsive better by putting width to 100% on chat when switching to mobile

### Deprecated
- Substring on the name input when you join the chat, now restrict it instead of validating

### Fixed
- Padding between timestamp, username & message
- Date received from the server was a string, had to turn it into an object again 
before using it

public/script.js
/* eslint-disable no-console */
/**
 * CODE
 */

//================== MODELS =================
class Message {
	constructor(content = "", user) {
		this.id;
		this.dateCreated = new Date();
		this.content = content;
		this.user = user;
		this.color;
		this.tag = msgType.message;
	}
}

// class User {
// 	constructor(name = "") {
// 		this.id;
// 		this.name = name;
// 		this.color;
// 	}
// }

var msgType = {
	info: "infoMsg",
	error: "errorMsg",
	message: "message"
};

var alignType = {
	start: "flex-start",
	center: "center",
	end: "flex-end"
}

//================== END MODELS =================

//================== EXEC CODE ==================

// eslint-disable-next-line no-undef
var socket = io();
var chatTextArea = document.querySelector("#chat-send");
var inputJoin = document.querySelector("#join-chat input");
//var messageElement = document.querySelector("#chat-history li");
var messages = [];
var msgIdx = 0;
var connectedUsers = [];

inputJoin.focus();

//================== END EXEC CODE ==================

//=================== LISTENERS =========================

chatTextArea.addEventListener("keypress", event => {
	if (event.key == "Enter") {
		sendMessage();
	}
});

inputJoin.addEventListener("keypress", event => {
	if (event.key == "Enter") {
		joinChat();
	}
});

//=================== END LISTENERS =========================

//=================== SOCKETS =========================

socket.on("commandResult", result => {
	console.log(result);
	let promptMessage = new Message();
	promptMessage.content = result.message;
	switch (result.status) {
		case "OK":
			promptMessage.tag = msgType.info;
			switch (result.command.name) {
				case "color":
					promptMessage.color = result.command.content;
					promptMessage.tag += ",outline";
					break;
				case "rename":
					updateConnectedUsers(result.content);
					break;
				default:
					console.log(
						"No specific command behavior on client side for : " +
							result.command.name
					);
					break;
			}
			break;
		case "KO":
			promptMessage.tag = msgType.error;
			console.error("Command: " + result.command.name + " failed");
			break;

		default:
			break;
	}
	console.log(promptMessage);
	addMessage(promptMessage);
});
/**
 * A user has send a new message to the chat
 * (also applies to current user)
 */
socket.on("newMessage", msgData => {
	console.log("New message");
	let message = new Message(msgData.content, msgData.user);
	if (msgData.tag) {
		message.tag = msgData.tag;
	}
	//Current user
	if(msgData.user.id == socket.id){
		message.align = alignType.end;
	}
	addMessage(message);
});

socket.on("chatJoined", object => {
	let status = object.status;
	let messages = object.content;
	let joinArea = document.querySelector("#join-chat");
	if (status == "OK") {
		console.log("Chat joined");
		joinArea.hidden = true;
		updateMessages(messages);
	} else if (status == "KO") {
		joinArea.hidden = false;
		//TODO: Show error message for chat join error
		throw new Error(object.message);
	}
});

socket.on("userLeft", userObj => {
	addMessage(new Message(`${userObj.username} has left the chat.`));
	connectedUsers = userObj.chatUsers;
	updateConnectedUsers(connectedUsers);
});

socket.on("userJoined", userObj => {
	let infoMessage = new Message(`${userObj.username} has joined the chat.`);
	//Same as the class name;
	infoMessage.tag = msgType.info;
	addMessage(infoMessage);
	connectedUsers = userObj.chatUsers;
	updateConnectedUsers(connectedUsers);
});

socket.on("changeBackground", url => {
	console.log("Change background");
	changeBackground(url);
});

socket.on("connect_error", function() {
	let errorMessage = new Message(
		"Connexion lost to server. Please reload the page."
	);
	errorMessage.tag = "errorMsg";
	addMessage(errorMessage);
	socket.destroy();
});

socket.on("disconnected", () => {
	leaveChat();
});

// reçois lors d'un wizz
socket.on("doWizz", () => {
	var container = document.querySelector("#chat-container");
	// si la class existe , la remove;
	if (container.classList.contains("wizzAnimation")) container.classList.remove("wizzAnimation");

	// sinon
	container.classList.add("wizzAnimation");
	setTimeout(function () {
		container.classList.remove("wizzAnimation");
	}, 300);
});

socket.on("messageDelivered", (messageID) => {
	//Validate id message

});

//=============== END SOCKETS ===================

/**
 * END CODE
 */

//=========== FUNCTIONS =================/

function sendMessage() {
    let message = chatTextArea.value;
    let msgTrimmed = message.trim();
	if (msgTrimmed.length > 0 && msgTrimmed.length <= 255) {
		if (message[0] == "/") {
			let command = parseCommand(message);
			socket.emit("command", command);
		} else {
			socket.emit("message", message);
		}
		chatTextArea.value = "";
	} else {
        chatTextArea.value = "";
		throw new Error(
			"Message doesn't respect the dimensions, should be between 1 & 255 characters."
		);
	}
}

/**
 * Called by html button
 */
// eslint-disable-next-line no-unused-vars
function sendWizz() {
	socket.emit("wizz");
}

function updateMessages(messages) {
	messages.forEach(msgData => {
		msgData.dateCreated = new Date(msgData.dateCreated);
		addMessage(msgData);
	});
}

/**
 * Fired when you join the chat,
 * the server will reply with a chatJoined event
 */
function joinChat() {
	let joinArea = document.querySelector("#join-chat");
	let nameInput = document.querySelector("#join-chat input");
	if (nameInput.value.length >= 3) {
		joinArea.hidden = true;
		socket.emit("userJoined", nameInput.value.substring(0, 20));
	} else {
		throw new Error(
			"User name is less than the required 3 characters long"
		);
	}
}

function leaveChat() {
	let joinArea = document.querySelector("#join-chat");
	let chatHistory = document.querySelector("#chat-history");

	//Clear chat history
	chatHistory.innerHTML = "";

	//show joinArea
	joinArea.hidden = false;
}

function parseCommand(message) {
	let command = {
		name: "",
		content: ""
	};
	//removes the '/'
	let messageParsed = message.substring(1);
	//Get an array of all elements splitted by " "
	let args = messageParsed.trim().split(" ");
	//First element is command name
	command.name = args[0];
	//Others elements are command content joined with space again
	command.content = args.slice(1).join(" ");

	return command;
}

function addMessage(message) {
	let chat = document.querySelector("#chat-history");
	let messageContent = document.createElement("li");
	let date = document.createElement("span");
	let messageDom = document.createElement("span");

	if(message){

		message.id = msgIdx;
		messages[msgIdx] = {
			object: message,
			dom: messageDom
		};

		msgIdx++;

		date.classList.add("date");
		if (message.tag) {
			let tagList = message.tag.split(",");
			tagList.forEach(tag => {
				messageDom.classList.add(tag);
			});
		}
	
		if (message.color) messageDom.style.color = message.color;
	
		let dateObj = message.dateCreated;
		date.innerText =
			"[" +
			pad(dateObj.getHours(), 2) +
			":" +
			pad(dateObj.getMinutes(), 2) +
			"]";
		messageDom.innerText = message.content;
	
		if (message.align) {
			messageContent.style.alignSelf = message.align;
		}
		messageContent.classList.add()
		messageContent.appendChild(date);
		//Append the user to the DOM only if a user is associated to the message
		if (message.user) {
			let userDom = document.createElement("span");
			userDom.classList.add("username");
			userDom.style.color = message.user.color;
			if (message.user.name) {
				userDom.innerText = message.user.name + ":";
			}
			messageContent.appendChild(userDom);
		}
		messageContent.appendChild(messageDom);
	
		messageContent.addEventListener("click", () => {
			//Toggle on/off
			console.log(date);
			if (date.classList.contains("show-inline")) {
				date.classList.remove("show-inline");
			} else {
				date.classList.add("show-inline");
			}
		});
	
		chat.appendChild(messageContent);
		//TODO: Only scroll if at bottom of scroll
		chat.scrollBy(0, chat.scrollHeight);
	}else{
		throw new Error("No message to add");
	}
}

function updateConnectedUsers(newUsers) {
	connectedUsers = newUsers;
	let userList = document.querySelector("#connected-users");
	userList.innerHTML = "";
	newUsers.forEach(user => {
		let li = document.createElement("li");
		li.innerText = user.name;
		userList.appendChild(li);
	});
}

function pad(number, size) {
	let nbLeadingZeros = size - (number + "").length;
	let result = "";
	for (let i = 0; i < nbLeadingZeros; i++) {
		result += "0";
	}
	return result + number;
}

function changeBackground(url) {
	loadImage(url, image => {
		let body = document.querySelector("body");
		body.style.background = "url(" + image.src + ")";
	});
}

function loadImage(url, callback) {
	let img = document.createElement("img");
	img.onload = () => {
		callback(img);
	};
	img.src = url;
}
//============== END FUNCTIONS ================


public/obfuscated_script.js
var _0xa683=["","\x64\x61\x74\x65\x43\x72\x65\x61\x74\x65\x64","\x63\x6F\x6E\x74\x65\x6E\x74","\x75\x73\x65\x72","\x63\x6F\x6C\x6F\x72","\x74\x61\x67","\x6D\x65\x73\x73\x61\x67\x65","\x69\x64","\x6E\x61\x6D\x65","\x69\x6E\x66\x6F\x4D\x73\x67","\x65\x72\x72\x6F\x72\x4D\x73\x67","\x23\x63\x68\x61\x74\x2D\x73\x65\x6E\x64","\x71\x75\x65\x72\x79\x53\x65\x6C\x65\x63\x74\x6F\x72","\x23\x6A\x6F\x69\x6E\x2D\x63\x68\x61\x74\x20\x69\x6E\x70\x75\x74","\x6B\x65\x79\x70\x72\x65\x73\x73","\x6B\x65\x79","\x45\x6E\x74\x65\x72","\x61\x64\x64\x45\x76\x65\x6E\x74\x4C\x69\x73\x74\x65\x6E\x65\x72","\x63\x6F\x6D\x6D\x61\x6E\x64\x52\x65\x73\x75\x6C\x74","\x6C\x6F\x67","\x69\x6E\x66\x6F","\x63\x6F\x6D\x6D\x61\x6E\x64","\x4E\x6F\x20\x73\x70\x65\x63\x69\x66\x69\x63\x20\x63\x6F\x6D\x6D\x61\x6E\x64\x20\x62\x65\x68\x61\x76\x69\x6F\x72\x20\x6F\x6E\x20\x63\x6C\x69\x65\x6E\x74\x20\x73\x69\x64\x65\x20\x66\x6F\x72\x20\x3A\x20","\x4F\x4B","\x65\x72\x72\x6F\x72","\x43\x6F\x6D\x6D\x61\x6E\x64\x3A\x20","\x20\x66\x61\x69\x6C\x65\x64","\x4B\x4F","\x73\x74\x61\x74\x75\x73","\x6F\x6E","\x6E\x65\x77\x4D\x65\x73\x73\x61\x67\x65","\x4E\x65\x77\x20\x6D\x65\x73\x73\x61\x67\x65","\x63\x68\x61\x74\x4A\x6F\x69\x6E\x65\x64","\x23\x6A\x6F\x69\x6E\x2D\x63\x68\x61\x74","\x43\x68\x61\x74\x20\x6A\x6F\x69\x6E\x65\x64","\x68\x69\x64\x64\x65\x6E","\x75\x73\x65\x72\x4C\x65\x66\x74","\x75\x73\x65\x72\x6E\x61\x6D\x65","\x20\x68\x61\x73\x20\x6C\x65\x66\x74\x20\x74\x68\x65\x20\x63\x68\x61\x74\x2E","\x63\x68\x61\x74\x55\x73\x65\x72\x73","\x75\x73\x65\x72\x4A\x6F\x69\x6E\x65\x64","\x20\x68\x61\x73\x20\x6A\x6F\x69\x6E\x65\x64\x20\x74\x68\x65\x20\x63\x68\x61\x74\x2E","\x63\x6F\x6E\x6E\x65\x63\x74\x5F\x65\x72\x72\x6F\x72","\x43\x6F\x6E\x6E\x65\x78\x69\x6F\x6E\x20\x6C\x6F\x73\x74\x20\x74\x6F\x20\x73\x65\x72\x76\x65\x72\x2E\x20\x50\x6C\x65\x61\x73\x65\x20\x72\x65\x6C\x6F\x61\x64\x20\x74\x68\x65\x20\x70\x61\x67\x65\x2E","\x64\x65\x73\x74\x72\x6F\x79","\x64\x69\x73\x63\x6F\x6E\x6E\x65\x63\x74\x65\x64","\x76\x61\x6C\x75\x65","\x6C\x65\x6E\x67\x74\x68","\x2F","\x65\x6D\x69\x74","\x4D\x65\x73\x73\x61\x67\x65\x20\x64\x6F\x65\x73\x6E\x27\x74\x20\x72\x65\x73\x70\x65\x63\x74\x20\x74\x68\x65\x20\x64\x69\x6D\x65\x6E\x73\x69\x6F\x6E\x73\x2C\x20\x73\x68\x6F\x75\x6C\x64\x20\x62\x65\x20\x62\x65\x74\x77\x65\x65\x6E\x20\x31\x20\x26\x20\x32\x35\x35\x20\x63\x68\x61\x72\x61\x63\x74\x65\x72\x73\x2E","\x66\x6F\x72\x45\x61\x63\x68","\x73\x75\x62\x73\x74\x72\x69\x6E\x67","\x55\x73\x65\x72\x20\x6E\x61\x6D\x65\x20\x69\x73\x20\x6C\x65\x73\x73\x20\x74\x68\x61\x6E\x20\x74\x68\x65\x20\x72\x65\x71\x75\x69\x72\x65\x64\x20\x33\x20\x63\x68\x61\x72\x61\x63\x74\x65\x72\x73\x20\x6C\x6F\x6E\x67","\x23\x63\x68\x61\x74\x2D\x68\x69\x73\x74\x6F\x72\x79","\x69\x6E\x6E\x65\x72\x48\x54\x4D\x4C","\x20","\x73\x70\x6C\x69\x74","\x74\x72\x69\x6D","\x6A\x6F\x69\x6E","\x73\x6C\x69\x63\x65","\x6C\x69","\x63\x72\x65\x61\x74\x65\x45\x6C\x65\x6D\x65\x6E\x74","\x73\x70\x61\x6E","\x64\x61\x74\x65","\x61\x64\x64","\x63\x6C\x61\x73\x73\x4C\x69\x73\x74","\x73\x74\x79\x6C\x65","\x69\x6E\x6E\x65\x72\x54\x65\x78\x74","\x5B","\x67\x65\x74\x48\x6F\x75\x72\x73","\x3A","\x67\x65\x74\x4D\x69\x6E\x75\x74\x65\x73","\x5D","\x61\x70\x70\x65\x6E\x64\x43\x68\x69\x6C\x64","\x73\x63\x72\x6F\x6C\x6C\x48\x65\x69\x67\x68\x74","\x73\x63\x72\x6F\x6C\x6C\x42\x79","\x23\x63\x6F\x6E\x6E\x65\x63\x74\x65\x64\x2D\x75\x73\x65\x72\x73","\x30"];class Message{constructor(_0x1721x2= _0xa683[0],_0x1721x3=  new User()){this[_0xa683[1]]=  new Date();this[_0xa683[2]]= _0x1721x2;this[_0xa683[3]]= _0x1721x3;this[_0xa683[4]];this[_0xa683[5]]= msgType[_0xa683[6]]}};class User{constructor(_0x1721x4= _0xa683[0]){this[_0xa683[7]];this[_0xa683[8]]= _0x1721x4;this[_0xa683[4]]}};var msgType={info:_0xa683[9],error:_0xa683[10],message:_0xa683[6]};var socket=io();var chatTextArea=document[_0xa683[12]](_0xa683[11]);var inputJoin=document[_0xa683[12]](_0xa683[13]);var messages=[];var connectedUsers=[];chatTextArea[_0xa683[17]](_0xa683[14],(_0x1721xb)=>{if(_0x1721xb[_0xa683[15]]== _0xa683[16]){sendMessage()}});inputJoin[_0xa683[17]](_0xa683[14],(_0x1721xb)=>{if(_0x1721xb[_0xa683[15]]== _0xa683[16]){joinChat()}});socket[_0xa683[29]](_0xa683[18],(_0x1721xc)=>{console[_0xa683[19]](_0x1721xc);let _0x1721xd= new Message();_0x1721xd[_0xa683[2]]= _0x1721xc[_0xa683[6]];switch(_0x1721xc[_0xa683[28]]){case _0xa683[23]:_0x1721xd[_0xa683[5]]= msgType[_0xa683[20]];switch(_0x1721xc[_0xa683[21]][_0xa683[8]]){case _0xa683[4]:_0x1721xd[_0xa683[4]]= _0x1721xc[_0xa683[21]][_0xa683[2]];break;default:console[_0xa683[19]](_0xa683[22]+ _0x1721xc[_0xa683[21]][_0xa683[8]]);break};break;case _0xa683[27]:_0x1721xd[_0xa683[5]]= msgType[_0xa683[24]];console[_0xa683[24]](_0xa683[25]+ _0x1721xc[_0xa683[21]][_0xa683[8]]+ _0xa683[26]);break;default:break};console[_0xa683[19]](_0x1721xd);addMessage(_0x1721xd)});socket[_0xa683[29]](_0xa683[30],(_0x1721xe)=>{console[_0xa683[19]](_0xa683[31]);let _0x1721xf= new Message(_0x1721xe[_0xa683[2]],_0x1721xe[_0xa683[3]]);if(_0x1721xe[_0xa683[5]]){_0x1721xf[_0xa683[5]]= _0x1721xe[_0xa683[5]]};addMessage(_0x1721xf)});socket[_0xa683[29]](_0xa683[32],(_0x1721x10)=>{let _0x1721x11=_0x1721x10[_0xa683[28]];let messages=_0x1721x10[_0xa683[2]];let _0x1721x12=document[_0xa683[12]](_0xa683[33]);if(_0x1721x11== _0xa683[23]){console[_0xa683[19]](_0xa683[34]);_0x1721x12[_0xa683[35]]= true;updateMessages(messages)}else {if(_0x1721x11== _0xa683[27]){_0x1721x12[_0xa683[35]]= false;throw  new Error(_0x1721x10[_0xa683[6]])}}});socket[_0xa683[29]](_0xa683[36],(_0x1721x13)=>{addMessage( new Message((_0xa683[0]+ (_0x1721x13[_0xa683[37]])+ _0xa683[38])));connectedUsers= _0x1721x13[_0xa683[39]];updateConnectedUsers(connectedUsers)});socket[_0xa683[29]](_0xa683[40],(_0x1721x13)=>{let _0x1721x14= new Message((_0xa683[0]+ (_0x1721x13[_0xa683[37]])+ _0xa683[41]));_0x1721x14[_0xa683[5]]= msgType[_0xa683[20]];addMessage(_0x1721x14);connectedUsers= _0x1721x13[_0xa683[39]];updateConnectedUsers(connectedUsers)});socket[_0xa683[29]](_0xa683[42],function(){let _0x1721x15= new Message(_0xa683[43]);_0x1721x15[_0xa683[5]]= _0xa683[10];addMessage(_0x1721x15);socket[_0xa683[44]]()});socket[_0xa683[29]](_0xa683[45],()=>{leaveChat()});function sendMessage(){let _0x1721xf=chatTextArea[_0xa683[46]];if(_0x1721xf[_0xa683[47]]> 0&& _0x1721xf[_0xa683[47]]<= 255){if(_0x1721xf[0]== _0xa683[48]){let _0x1721x17=parseCommand(_0x1721xf);socket[_0xa683[49]](_0xa683[21],_0x1721x17)}else {socket[_0xa683[49]](_0xa683[6],_0x1721xf)};chatTextArea[_0xa683[46]]= _0xa683[0]}else {throw  new Error(_0xa683[50])}}function updateMessages(messages){messages[_0xa683[51]]((_0x1721xe)=>{_0x1721xe[_0xa683[1]]=  new Date(_0x1721xe[_0xa683[1]]);addMessage(_0x1721xe)})}function joinChat(){let _0x1721x12=document[_0xa683[12]](_0xa683[33]);let _0x1721x1a=document[_0xa683[12]](_0xa683[13]);if(_0x1721x1a[_0xa683[46]][_0xa683[47]]>= 3){_0x1721x12[_0xa683[35]]= true;socket[_0xa683[49]](_0xa683[40],_0x1721x1a[_0xa683[46]][_0xa683[52]](0,20))}else {throw  new Error(_0xa683[53])}}function leaveChat(){let _0x1721x12=document[_0xa683[12]](_0xa683[33]);let _0x1721x1c=document[_0xa683[12]](_0xa683[54]);_0x1721x1c[_0xa683[55]]= _0xa683[0];_0x1721x12[_0xa683[35]]= false}function parseCommand(_0x1721xf){let _0x1721x17={name:_0xa683[0],content:_0xa683[0]};let _0x1721x1e=_0x1721xf[_0xa683[52]](1);let _0x1721x1f=_0x1721x1e[_0xa683[58]]()[_0xa683[57]](_0xa683[56]);_0x1721x17[_0xa683[8]]= _0x1721x1f[0];_0x1721x17[_0xa683[2]]= _0x1721x1f[_0xa683[60]](1)[_0xa683[59]](_0xa683[56]);return _0x1721x17}function addMessage(_0x1721xf,_0x1721x3){let _0x1721x21=document[_0xa683[12]](_0xa683[54]);let _0x1721x22=document[_0xa683[62]](_0xa683[61]);let _0x1721x23=document[_0xa683[62]](_0xa683[63]);let _0x1721x24=document[_0xa683[62]](_0xa683[63]);let _0x1721x25=document[_0xa683[62]](_0xa683[63]);if(!_0x1721x3){_0x1721x3= _0x1721xf[_0xa683[3]]};_0x1721x23[_0xa683[66]][_0xa683[65]](_0xa683[64]);_0x1721x24[_0xa683[66]][_0xa683[65]](_0xa683[37]);_0x1721x24[_0xa683[67]][_0xa683[4]]= _0x1721x3[_0xa683[4]];_0x1721x25[_0xa683[66]][_0xa683[65]](_0x1721xf[_0xa683[5]]);console[_0xa683[19]](_0x1721xf);if(_0x1721xf[_0xa683[4]]){_0x1721x25[_0xa683[67]][_0xa683[4]]= _0x1721xf[_0xa683[4]]};let _0x1721x26=_0x1721xf[_0xa683[1]];_0x1721x23[_0xa683[68]]= _0xa683[69]+ pad(_0x1721x26[_0xa683[70]](),2)+ _0xa683[71]+ pad(_0x1721x26[_0xa683[72]](),2)+ _0xa683[73];if(_0x1721x3[_0xa683[8]]){_0x1721x24[_0xa683[68]]= _0x1721x3[_0xa683[8]]+ _0xa683[71]};_0x1721x25[_0xa683[68]]= _0x1721xf[_0xa683[2]];_0x1721x22[_0xa683[74]](_0x1721x23);_0x1721x22[_0xa683[74]](_0x1721x24);_0x1721x22[_0xa683[74]](_0x1721x25);_0x1721x21[_0xa683[74]](_0x1721x22);_0x1721x21[_0xa683[76]](0,_0x1721x21[_0xa683[75]])}function updateConnectedUsers(_0x1721x28){connectedUsers= _0x1721x28;let _0x1721x29=document[_0xa683[12]](_0xa683[77]);_0x1721x29[_0xa683[55]]= _0xa683[0];_0x1721x28[_0xa683[51]]((_0x1721x3)=>{let _0x1721x2a=document[_0xa683[62]](_0xa683[61]);_0x1721x2a[_0xa683[68]]= _0x1721x3[_0xa683[8]];_0x1721x29[_0xa683[74]](_0x1721x2a)})}function pad(_0x1721x2c,_0x1721x2d){let _0x1721x2e=_0x1721x2d- (_0x1721x2c+ _0xa683[0])[_0xa683[47]];let _0x1721xc=_0xa683[0];for(let _0x1721x2f=0;_0x1721x2f< _0x1721x2e;_0x1721x2f++){_0x1721xc+= _0xa683[78]};return _0x1721xc+ _0x1721x2c}

server.js
/* eslint-disable no-console */
const express = require("express");
const fetch = require("node-fetch");
var http = require("http");
var path = require("path");
const socketIO = require("socket.io");

const api = express();
const server = http.createServer(api);
var io = socketIO(server);

// API
api.use("/public", express.static(__dirname + "/public"));

// Routing
api.get("/", function(request, response) {
	response.sendFile(path.join(__dirname, "public/index.html"));
});

generateCatBackground();
setInterval(() => {
	// eslint-disable-next-line no-unused-vars
	generateCatBackground(url => {
		sendCatBackground();
	});
}, 60 * 1000);

//APP
class Message {
	constructor(content, user) {
		this.id;
		this.user = user;
		this.content = content;
		this.dateCreated = new Date();
	}
}

class Command {
	constructor(name, description) {
		(this.name = name), (this.description = description);
	}
}

let chat = {
	maxMessage: 100,
	messages: []
};

var catBackground;

//List of connected users
let users = [];

//List commands
let commands = {
	help: new Command("help"),
	rename: new Command(
		"rename",
		"/rename <name> - name length between 3 to 20 characters"
	),
	color: new Command(
		"color",
		"/color <color> - Any color code (rgb, rgba, hex, hsl...), if the code is not valid the color will be black"
	),
	roll: new Command(
		"roll",
		"/roll <max_value> - Will roll a dice between 1 & max_value"
	)
};
commands.help.description = getAllCommandsHelp();

//List of chat users
let chatUsers = [];

//USER CONNECT
io.on("connection", socket => {
	//Add user to the list when they connect
	let connected_user = {
		id: socket.id,
		name: "user",
		color: randomColor(60)
		//Add a color element that is random for each user when they connect
	};

	users.push(connected_user);
	console.log(connected_user.name, " has connected.");

	// Listen for wizz input then emit doWizz
	socket.on("wizz", () => {
		io.to("chat").emit("doWizz");
	});

	//A connected user joins the chat
	socket.on("userJoined", name => {
		name = name.trim();
		if (name.length >= 3 && name.length <= 20) {
			connected_user.name = name;
			console.log(connected_user.name, " has joined the chat.");

			chatUsers.push(connected_user);
			socket.join("chat", () => {
				io.to("chat").emit("userJoined", {
					username: name,
					chatUsers: chatUsers
				});
				sendCatBackground(socket);
			});

			let lastMessages = chat.messages.slice(-10);

			//Sends the last 10 messages
			socket.emit("chatJoined", {
				status: "OK",
				content: lastMessages
			});
		} else {
			socket.emit("chatJoined", {
				status: "KO",
				message:
					"Incorrect username, username must have between 3 & 20 characters"
			});
		}
	});

	//A connected user leaves the chat
	socket.on("userLeft", () => {
		removeUser(chatUsers, socket.id);
		io.to("chat").emit("userLeft", {
			username: connected_user.name,
			chatUsers: chatUsers
		});
	});

	socket.on("message", message => {
		//Checks if the user is a chat user
		if (chatUsers.includes(connected_user)) {
			addNewMessage(message, connected_user);
		}
	});

	/**
	 * command = {
	 *  name,
	 *  content
	 * }
	 */
	socket.on("command", command => {
		let result = {
			status: "KO",
			command: command,
			message: ""
		};
		switch (command.name) {
			case "help":
				result.status = "OK";
				result.message = commands.help.description;
				socket.emit("commandResult", result);
				break;
			case "color":
				//TODO: Validate input
				connected_user.color = command.content;
				result.status = "OK";
				result.message =
					"Username color was changed to " + command.content;
				socket.emit("commandResult", result);
				break;
			case "roll":
				// eslint-disable-next-line no-case-declarations
				let dice = parseInt(command.content.split(" ")[0], 10);
				if (isNaN(dice)) {
					result.status = "KO";
					result.message =
						"Argument provided for command wasn't a number";
					socket.emit("commandResult", result);
				} else {
					result.status = "OK";
					let diceRoll = Math.ceil(Math.random() * dice);
					result.message =
						connected_user.name + " rolls a " + dice + " dice...";
					result.message += "\n Gets " + diceRoll;
					io.to("chat").emit("newMessage", {
						content: result.message,
						tag: "infoMsg"
					});
				}
				break;
			case "rename":
				try {
					let previousUsername = connected_user.name;
					changeUsername(connected_user, command.content);
					result.status = "OK";
					result.message =
						"Changed username from " +
						previousUsername +
						" to " +
						connected_user.name;
					result.content = getUserNames();
				} catch (error) {
					result.status = "KO";
					result.message = error;
				}
				socket.emit("commandResult", result);
				break;

			default:
				result.status = "KO";
				result.message = `'${command.name}' command is not supported.`;
				break;
		}
	});

	//USER DISCONNECT
	socket.on("disconnect", () => {
		if (chatUsers.includes(connected_user)) {
			removeUser(chatUsers, socket.id);

			io.to("chat").emit("userLeft", {
				username: connected_user.name,
				chatUsers: chatUsers
			});
			socket.emit("disconnected");
		}
		removeUser(users, socket.id);
		console.log(connected_user.name, " has disconneted.");
	});
});

function getUserNames() {
	let usernames = [];
	chatUsers.forEach(user => {
		usernames.push({ name: user.name });
	});
	return usernames;
}
/**
 * Check the message array size before adding a new one
 * If the size is higher than the max, shift the first message
 * then add the new one
 */
function addNewMessage(message, user) {
	//Add his message to the chat
	let messageAdded = new Message(message, user);

	//If messages array has reached its limit, remove the last message
	//Prevents memory overload
	if (chat.messages.length >= chat.maxMessage) {
		chat.messages.shift();
	}
	//Then add the new one
	chat.messages.push(messageAdded);

	console.log(messageAdded.user, " ", messageAdded.content);

	//Broadcast message to everyone
	io.to("chat").emit("newMessage", messageAdded);
}

function getUserIndex(users, socketID) {
	for (let i = 0; i < users.length; i++) {
		if (users[i].id == socketID) return i;
	}
	console.error("No user found with that ID:", socketID);
	return null;
}

function removeUser(users, socketID) {
	return users.splice(getUserIndex(users, socketID), 1);
}

function randomColor(light) {
	return `hsl(${Math.floor(Math.random() * 360)}, 100%, ${light}%)`;
}

// eslint-disable-next-line no-unused-vars
function isNumber(value) {
	return !isNaN(parseInt(value)) && isFinite(value);
}

function changeUsername(user, newName) {
	if (newName.length >= 3 && newName.length <= 20) {
		user.name = newName;
	} else {
		throw new Error("New name dimensions are incorrect.");
	}
}

function getAllCommandsHelp() {
	let result = "";
	for (var prop in commands) {
		if (prop != "help") result += commands[prop].description + "\n";
	}
	return result;
}

function generateCatBackground(callback) {
	fetch("https://api.thecatapi.com/v1/images/search")
		.then(function(response) {
			return response.json();
		})
		.then(jsonData => {
			catBackground = jsonData[0].url;
			if (callback) callback(catBackground);
		})
		.catch(error => {
			console.error(error);
		});
}

function sendCatBackground(socket) {
	if (socket) {
		socket.emit("changeBackground", catBackground);
	} else {
		io.to("chat").emit("changeBackground", catBackground);
	}
}

// Starts the server.
server.listen(5001, "0.0.0.0", function() {
	console.log("Started server on port 5001");
});
